server/src/main/java/org/opensearch/indices/IndicesService.java

```diff
@@ -998,13 +1006,33 @@ private EngineConfigFactory getEngineConfigFactory(final IndexSettings idxSettin
         return new EngineConfigFactory(this.pluginsService, idxSettings);
     }
 
+    private IngestionConsumerFactory getIngestionConsumerFactory(final IndexSettings idxSettings) {
+        final IndexMetadata indexMetadata = idxSettings.getIndexMetadata();
+        if (indexMetadata == null) {
+            return null;
+        }
+        if (indexMetadata.useIngestionSource()) {
+            String type = indexMetadata.getIngestionSource().getType().toUpperCase(Locale.ROOT);
+            if (!ingestionConsumerFactories.containsKey(type)) {
+                throw new IllegalArgumentException("No factory found for ingestion source type [" + type + "]");
+            }
+            return ingestionConsumerFactories.get(type);
+        }
+        return null;
+    }
+
     private EngineFactory getEngineFactory(final IndexSettings idxSettings) {
         final IndexMetadata indexMetadata = idxSettings.getIndexMetadata();
         if (indexMetadata != null && indexMetadata.getState() == IndexMetadata.State.CLOSE) {
             // NoOpEngine takes precedence as long as the index is closed
             return NoOpEngine::new;
         }
 
+        // streaming ingestion
+        if (indexMetadata != null && indexMetadata.useIngestionSource()) {
+            return new IngestionEngineFactory();
```

Comment:
I think we can add `IngestionEngineFactory` as a 2nd ctor param to `IngestionEngine` and pass through here, wdyt? Then we don't need to plumb it through IndexService/Shard.

Reply:
excellent suggestion. one less plumbing to worry about!